### PR TITLE
feat: add posix_rename and statvfs extensions

### PR DESCRIPTION
**Pull request by SamMousa**

### CODE REVIEW

**Comment by terrafrost** [comment]
I'll try to review this in the next few days.

Thanks!

**Review by terrafrost — COMMENTED** [review]
I like the changes (altho checkout my 2x inline comments) but, a couple of things:

1. Could you do these for the 3.0 branch? Adding new functions isn't a BC break and if they're in the master branch it could be years before these changes ever see the light of day if I'm being at all honest.

    Ideally you'd do 2x PR's - one for the 3.0 branch and one for the master branch wherein you merge your 3.0 based branch into your master based branch but if you only did 1x PR against the 3.0 branch I could take care of merging that to master, as well.
2. Add a DocBlock comment to `posix_rename` explaining that it's a wrapper for an OpenSSH specific SFTP extension and how it's different than `rename` (eg. [if i understand it correctly] if the destination file exists it'll overwrite the destination file vs failing) and add a DocBlock comment to `statvfs` explaining what that does (provides info on the filesystem vs on a specific file).

**Comment by terrafrost** [comment]
In reviewing some of the newer SFTP versions...  apparently SFTPv6 supports atomic (POSIX) style renames. [draft-ietf-secsh-filexfer-13](https://datatracker.ietf.org/doc/html/draft-ietf-secsh-filexfer-13#page-39) gives this explanation:

> If flags does not include SSH_FXP_RENAME_OVERWRITE, and there already exists a file with the name specified by newpath, the server MUST respond with SSH_FX_FILE_ALREADY_EXISTS.
> 
> If flags includes SSH_FXP_RENAME_ATOMIC, and the destination file already exists, it is replaced in an atomic fashion.  I.e., there is no observable instant in time where the name does not refer to either the old or the new file.  SSH_FXP_RENAME_ATOMIC implies SSH_FXP_RENAME_OVERWRITE.

So maybe building off of that a good explanation of what `posix_rename` does (for the docblock comments) would be something like this:

> Where rename() fails "if there already exists a file with the name specified by newpath" (draft-ietf-secsh-filexfer-02#section-6.5), posix_rename() overwrites the existing file in an atomic fashion. ie. "there is no observable instant in time where the name does not refer to either the old or the new file" (draft-ietf-secsh-filexfer-13#page-39).

And altho this function could get an SFTPv6 implementation that doesn't rely on OpenSSH-specific extensions that's something I can do on my own.

File: phpseclib/Net/SFTP.php [hunk 1]
```diff
@@ -708,7 +717,7 @@ public function chdir(string $dir): bool
         // assume current dir if $dir is empty
         if ($dir === '') {
             $dir = './';
-        // suffix a slash if needed
```
[the remark this answers is not in the document]
Comment (SamMousa) [reply phpseclib/Net/SFTP.php:711]: I don't disagree, but this seems like something php-cs-fixer should fix. Looking at master it's currently not working there either.

**Comment by SamMousa** [comment]
I changed the 2 you mentioned explicitly. The other changes come from php-cs-fixer which I took from CI (currently failing on master).

**Comment by terrafrost** [comment]
This has been merged to master and the non-CS changes have been backported to the 3.0 branch:

https://github.com/phpseclib/phpseclib/commit/ce0aeec4023bb984b497ab1e97cd258e84b12fee

In the course of doing that I had to make a few other changes to what you did for the master branch.

- I renamed `\RuntimeException` to `RuntimeException` and `\UnexpectedValueException` to `UnexpectedValueException`.  In the master branch this was done with PR #1835 . The justifiation can be found in https://github.com/phpseclib/phpseclib/pull/1827#discussion_r947954270 :

  > I believe anything thrown by this library should implement this interface (so yes). This is how Symfony does it.

- I added the following to the start of `statvfs()` (it was already present in `posix_rename()`):

   ```php
        if (!$this->precheck()) {
            return false;
        }
   ```

I'll try to review your other PR this coming week.

Thanks!

**Pull request merged**